neo_force_scheme/utils.py

```diff
@@ -45,41 +43,62 @@ def pickle_save_matrix(filename, distance_matrix, size):
     size = len(projection)
     total = len(distance_matrix)
     error = 0
 
-    for ins2 in numba.prange(size):
-        if ins1 != ins2:
-            x1x2 = projection[ins2][0] - projection[ins1][0]
-            y1y2 = projection[ins2][1] - projection[ins1][1]
-            dr2 = max(math.sqrt(x1x2 * x1x2 + y1y2 * y1y2), 0.0001)
-
-            # getting te index in the distance matrix and getting the value
-            r = (ins1 + ins2 - math.fabs(ins1 - ins2)) / 2  # min(i,j)
-            s = (ins1 + ins2 + math.fabs(ins1 - ins2)) / 2  # max(i,j)
-            drn = distance_matrix[int(total - ((size - r) * (size - r + 1) / 2) + (s - r))]
-
-            # calculate the movement
-            delta = (drn - dr2)
-            error += math.fabs(delta)
-
-            # moving
-            projection[ins2][0] += learning_rate * delta * (x1x2 / dr2)
-            projection[ins2][1] += learning_rate * delta * (y1y2 / dr2)
+    if n_dimension == 2:
+        for ins2 in numba.prange(size):
+            if ins1 != ins2:
+                x1x2 = projection[ins2][0] - projection[ins1][0]
+                y1y2 = projection[ins2][1] - projection[ins1][1]
+                dr2 = max(math.sqrt(x1x2 * x1x2 + y1y2 * y1y2), 0.0001)
+
+                # getting te index in the distance matrix and getting the value
+                r = (ins1 + ins2 - math.fabs(ins1 - ins2)) / 2  # min(i,j,k)
+                s = (ins1 + ins2 + math.fabs(ins1 - ins2)) / 2  # max(i,j,k)
+                drn = distance_matrix[int(total - ((size - r) * (size - r + 1) / 2) + (s - r))]
+
+                # calculate the movement
+                delta = (drn - dr2)
+                error += math.fabs(delta)
+
+                projection[ins2][0] += learning_rate * delta * (x1x2 / dr2)
+                projection[ins2][1] += learning_rate * delta * (y1y2 / dr2)
+
+    elif n_dimension == 3:
+        for ins2 in numba.prange(size):
+            if ins1 != ins2:
+                x1x2 = projection[ins2][0] - projection[ins1][0]
+                y1y2 = projection[ins2][1] - projection[ins1][1]
+                z1z2 = projection[ins2][2] - projection[ins1][2]
+                dr2 = max(math.sqrt(x1x2 * x1x2 + y1y2 * y1y2 + z1z2 * z1z2), 0.0001)
+
+                # getting te index in the distance matrix and getting the value
+                r = (ins1 + ins2 - math.fabs(ins1 - ins2)) / 2  # min(i,j,k)
+                s = (ins1 + ins2 + math.fabs(ins1 - ins2)) / 2  # max(i,j,k)
+                drn = distance_matrix[int(total - ((size - r) * (size - r + 1) / 2) + (s - r))]
+
+                # calculate the movement
+                delta = (drn - dr2)
+                error += math.fabs(delta)
+
+                projection[ins2][0] += learning_rate * delta * (x1x2 / dr2)
+                projection[ins2][1] += learning_rate * delta * (y1y2 / dr2)
+                projection[ins2][2] += learning_rate * delta * (z1z2 / dr2)
```

Comment:
for now, this is fine, but can you also add a raise NotImplementedException if n_dimensions is >3? I think we (by that I mean our lab, not necessarily during your project) will end up generalizing this in the future.